File: ios/Classes/BactrackFlutterPlugin.swift
import Flutter
import UIKit
import BreathalyzerSDK

public class BactrackFlutterPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var bacTrackAPI: BacTrackAPI?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "bactrack_flutter", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "bactrack_flutter/events", binaryMessenger: registrar.messenger())
        
        let instance = BactrackFlutterPlugin()
        instance.channel = channel
        instance.eventChannel = eventChannel
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            guard let args = call.arguments as? [String: Any],
                  let apiKey = args["apiKey"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "API key required", details: nil))
                return
            }
            initializeSDK(apiKey: apiKey, result: result)
            
        case "connectToNearestBreathalyzer":
            connectToNearestBreathalyzer(result: result)
            
        case "disconnect":
            disconnect(result: result)
            
        case "startCountdown":
            startCountdown(result: result)
            
        case "getBatteryVoltage":
            getBatteryVoltage(result: result)
            
        case "getSerialNumber":
            getSerialNumber(result: result)
            
        case "getFirmwareVersion":
            // SDK doesn't have this method - return not supported
            result(nil)
            
        case "getUseCount":
            // SDK doesn't have this method - return not supported
            result(nil)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - SDK Methods
    
    private func initializeSDK(apiKey: String, result: @escaping FlutterResult) {
        bacTrackAPI = BacTrackAPI(delegate: self, andAPIKey: apiKey)
        result(true)
    }
    
    private func connectToNearestBreathalyzer(result: @escaping FlutterResult) {
        guard let api = bacTrackAPI else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "SDK not initialized", details: nil))
            return
        }
        api.connectToNearestBreathalyzer()
        result(nil)
    }
    
    private func disconnect(result: @escaping FlutterResult) {
        bacTrackAPI?.disconnect()
        result(nil)
    }
    
    private func startCountdown(result: @escaping FlutterResult) {
        guard let api = bacTrackAPI else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "SDK not initialized", details: nil))
            return
        }
        let success = api.startCountdown()
        result(success)
    }
    
    private func getBatteryVoltage(result: @escaping FlutterResult) {
        bacTrackAPI?.getBreathalyzerBatteryLevel()
        result(nil)
    }
    
    private func getSerialNumber(result: @escaping FlutterResult) {
        bacTrackAPI?.getBreathalyzerSerialNumber()
        result(nil)
    }
    
    // MARK: - Helper
    
    private func sendEvent(state: String, message: String? = nil) {
        var event: [String: Any] = ["state": state]
        if let msg = message {
            event["message"] = msg
        }
        eventSink?(event)
    }
}

// MARK: - FlutterStreamHandler

extension BactrackFlutterPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

// MARK: - BacTrackAPIDelegate

extension BactrackFlutterPlugin: BacTrackAPIDelegate {
    
    // Required delegate method
    public func bacTrackAPIKeyDeclined(_ errorMessage: String!) {
        sendEvent(state: "apiKeyDeclined", message: errorMessage ?? "Unknown error")
    }
    
    // Optional delegate methods
    public func bacTrackAPIKeyAuthorized() {
        sendEvent(state: "apiKeyAuthorized")
    }
    
    public func bacTrackConnected(_ device: BACtrackDeviceType) {
        sendEvent(state: "connected", message: String(device.rawValue))
    }
    
    public func bacTrackDisconnected() {
        sendEvent(state: "disconnected")
    }
    
    public func bacTrackConnectTimeout() {
        sendEvent(state: "connectionTimeout")
    }
    
    public func bacTrackCountdown(_ seconds: NSNumber!, executionFailure error: Bool) {
        if error {
            sendEvent(state: "error", message: "Countdown execution failure")
        } else {
            sendEvent(state: "countdown", message: seconds?.stringValue ?? "0")
        }
    }
    
    public func bacTrackStart() {
        sendEvent(state: "start")
    }
    
    public func bacTrackBlow(_ breathFractionRemaining: NSNumber!) {
        sendEvent(state: "blow", message: breathFractionRemaining?.stringValue)
    }
    
    public func bacTrackAnalyzing() {
        sendEvent(state: "analyzing")
    }
    
    public func bacTrackResults(_ bac: CGFloat) {
        sendEvent(state: "result", message: String(format: "%.4f", bac))
    }
    
    public func bacTrackError(_ error: Error!) {
        sendEvent(state: "error", message: error?.localizedDescription ?? "Unknown error")
    }
    
    public func bacTrackBatteryLevel(_ number: NSNumber!) {
        sendEvent(state: "batteryLevel", message: number?.stringValue ?? "0")
    }
    
    public func bacTrackSerial(_ serial_hex: String!) {
        sendEvent(state: "serialNumber", message: serial_hex ?? "")
    }
    
    public func bacTrackFoundBreathalyzer(_ breathalyzer: Breathalyzer!) {
        sendEvent(state: "foundBreathalyzer", message: breathalyzer?.uuid ?? "")
    }
}

File: android/src/main/kotlin/com/wellcentiv/bactrack_flutter/BactrackFlutterPlugin.kt
package com.wellcentiv.bactrack_flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import BreathalyzerSDK.API.BACtrackAPI
import BreathalyzerSDK.API.BACtrackAPICallbacks
import BreathalyzerSDK.Constants.BACTrackDeviceType
import BreathalyzerSDK.Constants.BACtrackUnit
import BreathalyzerSDK.Exceptions.BluetoothLENotSupportedException
import BreathalyzerSDK.Exceptions.BluetoothNotEnabledException
import BreathalyzerSDK.Exceptions.LocationServicesNotEnabledException

/** BactrackFlutterPlugin */
class BactrackFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, EventChannel.StreamHandler {
    companion object {
        private const val TAG = "BactrackFlutter"
    }

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var context: Context? = null
    private var bactrackAPI: BACtrackAPI? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "bactrack_flutter")
        channel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "bactrack_flutter/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> {
                val apiKey = call.argument<String>("apiKey")
                if (apiKey == null) {
                    result.error("INVALID_ARGS", "API key required", null)
                    return
                }
                initializeSDK(apiKey, result)
            }
            "connectToNearestBreathalyzer" -> {
                val timeoutMs = call.argument<Int>("timeoutMs")
                connectToNearestBreathalyzer(timeoutMs, result)
            }
            "disconnect" -> {
                disconnect(result)
            }
            "startCountdown" -> {
                startCountdown(result)
            }
            "getBatteryVoltage" -> {
                getBatteryVoltage(result)
            }
            "getSerialNumber" -> {
                getSerialNumber(result)
            }
            "getFirmwareVersion" -> {
                getFirmwareVersion(result)
            }
            "getUseCount" -> {
                getUseCount(result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    // SDK Methods
    
    private fun initializeSDK(apiKey: String, result: Result) {
        val ctx = context
        if (ctx == null) {
            result.error("NO_CONTEXT", "Application context not available", null)
            return
        }
        
        try {
            bactrackAPI = BACtrackAPI(ctx, bactrackCallbacks, apiKey)
            result.success(true)
        } catch (e: BluetoothLENotSupportedException) {
            sendEvent("bluetoothNotAvailable", "Bluetooth LE not supported")
            result.success(false)
        } catch (e: BluetoothNotEnabledException) {
            sendEvent("bluetoothOff", "Bluetooth is not enabled")
            result.success(false)
        } catch (e: LocationServicesNotEnabledException) {
            sendEvent("error", "Location services not enabled")
            result.success(false)
        } catch (e: Exception) {
            sendEvent("error", e.message ?: "Unknown error")
            result.success(false)
        }
    }

    private fun connectToNearestBreathalyzer(timeoutMs: Int?, result: Result) {
        val api = bactrackAPI
        if (api == null) {
            result.error("NOT_INITIALIZED", "SDK not initialized", null)
            return
        }
        
        api.connectToNearestBreathalyzer()
        result.success(null)
    }

    private fun disconnect(result: Result) {
        bactrackAPI?.disconnect()
        result.success(null)
    }

    private fun startCountdown(result: Result) {
        val api = bactrackAPI
        if (api == null) {
            result.error("NOT_INITIALIZED", "SDK not initialized", null)
            return
        }
        api.startCountdown()
        result.success(true)
    }

    private fun getBatteryVoltage(result: Result) {
        bactrackAPI?.getBreathalyzerBatteryVoltage()
        result.success(null)
    }

    private fun getSerialNumber(result: Result) {
        bactrackAPI?.getSerialNumber()
        result.success(null)
    }

    private fun getFirmwareVersion(result: Result) {
        bactrackAPI?.getFirmwareVersion()
        result.success(null)
    }

    private fun getUseCount(result: Result) {
        bactrackAPI?.getUseCount()
        result.success(null)
    }

    // Helper
    
    private fun sendEvent(state: String, message: String? = null) {
        val event = mutableMapOf<String, Any>("state" to state)
        message?.let { event["message"] = it }
        // Must dispatch to main thread for Flutter platform channel
        mainHandler.post {
            eventSink?.success(event)
        }
    }

    // BACtrack Callbacks
    
    private val bactrackCallbacks = object : BACtrackAPICallbacks {
        override fun BACtrackAPIKeyDeclined(errorMessage: String?) {
            sendEvent("apiKeyDeclined", errorMessage)
        }

        override fun BACtrackAPIKeyAuthorized() {
            sendEvent("apiKeyAuthorized")
        }

        override fun BACtrackConnected(deviceType: BACTrackDeviceType?) {
            sendEvent("connected", deviceType?.toString())
        }

        override fun BACtrackDidConnect(message: String?) {
            sendEvent("didConnect", message)
        }

        override fun BACtrackDisconnected() {
            sendEvent("disconnected")
        }

        override fun BACtrackConnectionTimeout() {
            sendEvent("connectionTimeout")
        }

        override fun BACtrackFoundBreathalyzer(device: BACtrackAPI.BACtrackDevice?) {
            sendEvent("foundBreathalyzer", device?.toString())
        }

        override fun BACtrackCountdown(currentCountdownCount: Int) {
            sendEvent("countdown", currentCountdownCount.toString())
        }

        override fun BACtrackStart() {
            sendEvent("start")
        }

        override fun BACtrackBlow(blowProgress: Float) {
            sendEvent("blow", blowProgress.toString())
        }

        override fun BACtrackAnalyzing() {
            sendEvent("analyzing")
        }

        override fun BACtrackResults(measuredBac: Float) {
            sendEvent("result", String.format("%.4f", measuredBac))
        }

        override fun BACtrackFirmwareVersion(version: String?) {
            sendEvent("firmwareVersion", version)
        }

        override fun BACtrackSerial(serialNumber: String?) {
            sendEvent("serialNumber", serialNumber)
        }

        override fun BACtrackBatteryVoltage(voltage: Float) {
            sendEvent("batteryVoltage", voltage.toString())
        }

        override fun BACtrackBatteryLevel(level: Int) {
            sendEvent("batteryLevel", level.toString())
        }

        override fun BACtrackUseCount(useCount: Int) {
            sendEvent("useCount", useCount.toString())
        }

        override fun BACtrackError(errorCode: Int) {
            sendEvent("error", "Error code: $errorCode")
        }

        override fun BACtrackUnits(units: BACtrackUnit?) {
            val unitsString = when (units) {
                BACtrackUnit.BACtrackUnit_bac -> "BAC%"
                BACtrackUnit.BACtrackUnit_permille -> "permille"
                BACtrackUnit.BACtrackUnit_mgL -> "mg/L"
                BACtrackUnit.BACtrackUnit_mg -> "mg/dL"
                BACtrackUnit.BACtrackUnit_permilleByMass -> "permille"
                else -> "unknown"
            }
            sendEvent("units", unitsString)
        }
    }

    // EventChannel.StreamHandler
    
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ActivityAware
    
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {}
    override fun onDetachedFromActivityForConfigChanges() {}
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}
    override fun onDetachedFromActivity() {}

    // FlutterPlugin
    
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        bactrackAPI?.disconnect()
        bactrackAPI = null
        context = null
    }
}
